Crossgen2 fixes to enable composite build with shared framework

trylek opened the pull request:
These changes involve targeted fixes to issues seen due to the new
build mode and generic signature encoding fixes identified in
my offline investigation with JanV. While I locally see the tests
passing with just a handful of failures, I still seem unable to
make the tests run in the lab; I keep investigating that but
I believe it's useful to merge this delta in to let us parallelize
follow-up efforts.

Thanks

Tomas

trylek committed:
Crossgen2 fixes to enable composite build with shared framework

These changes involve targeted fixes to issues seen due to the new
build mode and generic signature encoding fixes identified in
my offline investigation with JanV. While I locally see the tests
passing with just a handful of failures, I still seem unable to
make the tests run in the lab; I keep investigating that but
I believe it's useful to merge this delta in to let us parallelize
follow-up efforts.

Thanks

Tomas

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InliningInfoNode.cs:
```diff
@@ -66,6 +66,13 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
 
+                    if (!factory.CompilationModuleGroup.VersionsWithMethodBody(inlinee))
+                    {
+                        // We cannot record inlining info across version bubble as cross-bubble assemblies
```
We should assert that this only happens for NonVersionable methods. (If its happening for any other method, we shouldn't actually be inclining)

trylek replied:
Good point, fixed in 3rd commit. Could you please take another look when you have a chance?

davidwrighton reviewed (changes requested):
One minor nit, but otherwise looks good.

MichalStrehovsky commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs:
```diff
@@ -310,6 +310,12 @@ public void EmitTypeSignature(TypeDesc typeDesc, SignatureContext context)
                 case TypeFlags.ValueType:
                 case TypeFlags.Nullable:
                 case TypeFlags.Enum:
+                    if (typeDesc == PrimitiveTypeProvider.GetPrimitiveType(typeDesc.Context, PrimitiveTypeCode.TypedReference))
```
Nit: this is more compact/readable (same compact form can be used for the diff in SignatureContext.cs)

```suggestion
                    if (typeDesc.IsWellKnownType(WellKnownType.TypedReference))
```

trylek replied:
Thanks for pointing that out, fixed in 2nd commit along with one more place that was using the same check.

trylek committed:
Simplify check for TypedReference per Michal's PR feedback

trylek committed:
Inlining check per DavidWr's PR feedback

Based on David's suggestion I have added an assertion to check that
cross-bubble inlining only takes place when the inlinee is marked
as non-versionable. To avoid duplicating the custom attribute
check I have added a new helper method IsNonVersionable to TypeExtensions
and MethodExtensions.

Thanks

Tomas

davidwrighton reviewed (approved):
Looks good

trylek commented:
Hit two unrelated bugs, filed issues for both:

1) Error in BinderTracingTest.Basic on all OS / architecture combos in Crossgen2 compilation mode: 

https://github.com/dotnet/runtime/issues/34509

2) Infra failure in perf test (error in CI setup script on Linux x64):

https://github.com/dotnet/runtime/issues/34511
